Say "Invitation Sent" if only one invitation was sent, and add multiple UUID invites

Ozzystrasza opened the pull request:
Fixes https://github.com/HabitRPG/habitrpg/issues/7789
### Changes

Added the string "invitationSent": "Invitation sent!" in the groups.json file

In the inviteToGroupCtrl.js file, added a test to check if more than 1 invite is being sent to choose which string will be used in the notification, plural or singular.

In the invite-friends.jade file, removed //- on the already existing code to enable multiple UUID invites for party and guild..

Tested locally using three accounts, worked on the following tests for both party and guild:
Inviting one by UUID;
Inviting two by UUID;
Inviting one by email;
Inviting two by email.

![2016-07-16](https://cloud.githubusercontent.com/assets/20071290/16894509/b2bda4de-4b2e-11e6-8956-d4186afc7fa9.png)

![2016-07-16 1](https://cloud.githubusercontent.com/assets/20071290/16929445/395f30d0-4d0e-11e6-975b-c39517a354ef.png)

![2016-07-18](https://cloud.githubusercontent.com/assets/20071290/16929469/4c10a510-4d0e-11e6-9e32-59c3f256a778.png)

---

UUID: 8389891a-ba52-4580-b08b-2cb706c7e0b4

Ozzystrasza committed:
Add a string for "Invitation Sent"

Ozzystrasza committed:
Add invitations test

Test if more than 1 invitation is being sent, to use the notification text in plural or singular.

Ozzystrasza committed:
Add plus button

Removed comment tags (//-) to enable multiple UUID invites for party and guild.

khipkin commented on website/client/js/controllers/inviteToGroupCtrl.js:
```diff
@@ -44,8 +44,13 @@ habitrpg.controller('InviteToGroupCtrl', ['$scope', '$rootScope', 'User', 'Group
         .then(function() {
-          Notification.text(window.env.t('invitationsSent'));
-          _resetInvitees();
+           if ((inviteMethod === 'email' && invitationDetails.emails.length > 1) || (inviteMethod === 'uuid' && invitationDetails.uuids.length > 1)) { // test if it's inviting more than 1 emails or uuid
```
If either `inviteMethod === 'email'` or `inviteMethod === 'uuid'` must be true if either `invitationDetails` has a length > 1, then it's unnecessary to check.

`if (invitationDetails.emails.length + invitationDetails.uuids.length > 1) {`

Ozzystrasza replied:
@Hkly Did what you said, and it worked fine. Sorry it took so long

khipkin commented on website/views/shared/modals/invite-friends.jade:
```diff
@@ -12,11 +12,10 @@ script(type='text/ng-template', id='modals/invite-guild.html')
           tr(ng-repeat='user in invitees track by $index')
             td
               input.form-control(type='text', ng-model='user.uuid')
-          //- @TODO: Refactor src/controllers/groups api.invite to use promises before implimenting this
```
Are you sure you can uncomment this?

khipkin replied:
Nevermind!

khipkin commented on website/views/shared/modals/invite-friends.jade:
```diff
@@ -91,11 +90,10 @@ script(type='text/ng-template', id='modals/invite-party.html')
               tr(ng-repeat='user in invitees track by $index')
                 td
                   input.form-control(type='text', ng-model='user.uuid')
-              //- @TODO: Refactor src/controllers/groups api.invite to use promises before implimenting this
```
Are you sure you can uncomment this?

khipkin replied:
Nevermind!

khipkin commented:
Needs tests.

Ozzystrasza commented on website/client/js/controllers/inviteToGroupCtrl.js:
```diff
@@ -44,8 +44,13 @@ habitrpg.controller('InviteToGroupCtrl', ['$scope', '$rootScope', 'User', 'Group
         .then(function() {
-          Notification.text(window.env.t('invitationsSent'));
-          _resetInvitees();
+           if ((inviteMethod === 'email' && invitationDetails.emails.length > 1) || (inviteMethod === 'uuid' && invitationDetails.uuids.length > 1)) { // test if it's inviting more than 1 emails or uuid
```
Ohh ok, I'm not very familiar with this language so I didn't know you could do that! Thanks! I will change it.

Ozzystrasza replied:
@khipkin Are you sure that is correct? I made that change locally and the invite button stopped working.
I made those tests because if `inviteMethod === 'email'` then `invitationDetails.uuids` is not supposed to exist, same for `inviteMethod === 'uuid'` and `invitationDetails.emails`.

khipkin replied:
@Ozzystrasza Oh okay, then what you did is perfect! :+1: Sorry for the confusion.

Hkly replied:
> `inviteMethod === 'email'` then `invitationDetails.uuids` is not supposed to exist

~~What about just checking if either exists then?~~ My bad, this doesn't account for single emails/uuids haha

``` javascript
if (invitationDetails.emails || invitationDetails.uuids) {
```

Or for slightly better readability without changing your logic, you could do something like

``` javascript
var isMultipleEmailInvites = inviteMethod === 'email' && invitationDetails.emails.length > 1;
var isMultipleUUIDInvites = inviteMethod === 'uuid' && invitationDetails.uuids.length > 1;

if (isMultipleEmailInvites || isMultipleUUIDInvites) {
// etc.
}
```

Edit:
Or maybe this with a note indicating that either property only exists at a time.

``` javascript
var invitesSent = invitationDetails.emails || invitationDetails.uuids;

if (invitesSent > 1) {
// etc.
}
```

Ozzystrasza commented:
Added some screen shots of some tests I made, is that enough? Or should I do something more?

khipkin commented on website/client/js/controllers/inviteToGroupCtrl.js:
```diff
@@ -44,8 +44,13 @@ habitrpg.controller('InviteToGroupCtrl', ['$scope', '$rootScope', 'User', 'Group
+            else {
+                Notification.text(window.env.t('invitationSent'));
+            }
+            _resetInvitees();
```
Add tests to [`inviteToGroupCtrlSpec.js`](https://github.com/Ozzystrasza/habitrpg/blob/18d975469328fff7ac2798d32e70498e7c0eb57f/test/spec/controllers/inviteToGroupCtrlSpec.js) that test these changes.

khipkin replied:
@Ozzystrasza You still have to add tests to `inviteToGroupCtrlSpec.js`, as I mentioned here.

Ozzystrasza replied:
@khipkin  Could you give me an example of how those tests look like? If it's not too much trouble

khipkin replied:
@Ozzystrasza Take a look at the existing tests in [inviteToGroupCtrlSpec.js](https://github.com/Ozzystrasza/habitrpg/blob/934aa2f04e7d690f734689844f217141ecddc1b7/test/spec/controllers/inviteToGroupCtrlSpec.js).

Ozzystrasza replied:
@khipkin I'm so sorry! This whole time I thought you were talking about the `inviteToGroupCtrl.js` file, I didn't see the spec at the end. 
Ok, so I checked it out, and it seems that the multiple uuid invites part is already covered (unless I'm mistaken). But I still couldn't figure out how to test the changes I made to the `inviteToGroupCtrl.js` file. All I did here was change the string of the notification, I couldn't find anything like that in the `inviteToGroupCtrlSpec.js`. I'm sorry :(

Ozzystrasza commented:
Added the tests for the changes in `inviteToGroupCtrlSpec.js`. But I noticed something that might be an issue, and if it is I believe it is relevant here.  When you click on the "Send Invitations" button even though you left the "Email" or "User ID" fields blank, it still says "Invitation Sent" (This was not caused by my changes, I tested on the main website and it happens as well). 
So I was thinking, maybe create a new string like "Email or User ID can't be blank" or something like that, and add another test before sending the notification. 
Unless you guys don't think it is an issue.

khipkin commented:
@Ozzystrasza Can you push your change(s) with the tests so we can look at them?  And maybe create a new issue for the bug you found so it can be triaged.  Then, if the repo maintainers want you to fix it, they'll tell you.

Ozzystrasza commented:
@khipkin I definitely can, but unfortunately I'm still very inexperienced in all this and I don't know how to do that. Can you help me?

Ozzystrasza commented:
@khipkin I copied the changes to the develop branch on my fork of the project, is that what you wanted? I'm sorry for this, I really don't know if this is the right place to be asking for help.

crookedneighbor commented:
You would need to copy them over to the branch you used to open the pull request. Looks like you've named that patch-3. You would then push up the changes to Githib.

Ozzystrasza commented:
Actually I edited the files directly from the Github website, and that branch patch-3 was created when I commited the changes. I did that after testing the same changes locally, but I'm starting to think that wasn't a good way to do this.

khipkin commented:
@Ozzystrasza This PR is a request to merge your commits on your `patch-3` branch into the main repo's `develop` branch.  If your other relevant commits aren't on the `patch-3` branch of your fork, they won't show up on this PR.

Are you sure you're on the `patch-3` branch? Are you sure you pushed your local changes and didn't just commit them?
Type `git status` and tell us the exact output so we can try to figure out the answers to those questions.

Ozzystrasza commented:
@khipkin First of all you definitely have been more than a random contributor, your help has been invaluable to me.
That being said. I never created a new branch locally, I just edited the files and tested on localhost:3000. Once I saw it was working fine, I went to github.com on my fork of habitrpg and edited the same files with the same changes, and then choose the option "Commit into a new branch" thus creating `patch-3`. After I finished all the changes I clicked on pull request. 

Here is the output on the `git status` command:

```
 MINGW64 ~/documents/habitica/habitrpg (develop)
$ git status
On branch develop
Your branch is up-to-date with 'origin/develop'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   common/locales/en/generic.json
        modified:   common/locales/en/groups.json
        modified:   website/client/js/controllers/inviteToGroupCtrl.js
        modified:   website/server/models/user/schema.js
        modified:   website/views/shared/header/menu.jade
        modified:   website/views/shared/modals/invite-friends.jade

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        common/audio/dewinTheme/

no changes added to commit (use "git add" and/or "git commit -a")
```

It shows some other files not relevant to this PR cause I have another PR pending.

If I do a `git push` here is the output:

```
 MINGW64 ~/documents/habitica/habitrpg (develop)
$ git push
To https://github.com/Ozzystrasza/habitrpg.git
 ! [rejected]        develop -> develop (fetch first)
error: failed to push some refs to 'https://github.com/Ozzystrasza/habitrpg.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

khipkin commented:
@Ozzystrasza From now on, you should follow the recommended Git workflow from [the Habitica Wiki](http://habitica.wikia.com/wiki/Using_Habitica_Git).  Having a dedicated branch for each PR will make your life a lot easier, allowing you to create and update multiple PRs at the same time while keeping their changes separate.

---

Let's go over some Git concepts:
- Your Habitica fork on Github is a copy of the main Habitica repo.  You keep your fork and the original repo up-to-date with each other by rebasing (as described in the wiki link above) and submitting PRs like this one.
- Your local repo is a copy of your Habitica fork on Github.  You keep your local repo and your fork up-to-date with each other by pushing and pulling.
- If you edit files on the Github website, like you did when you created the `patch-3` branch, then your fork has changes that your local repo doesn't know about.  To update your local repo, you need to `git pull`.
- If you edit files on your computer and commit, then your local repo has changes that your fork doesn't know about.  To update your fork, you need to `git push`.
- Each repo has a history, which is basically a tree of commits, or a flowchart of changes in order.  When you create a new branch, you also create a new "branch" in this flowchart.  Merging brings the branches back together and consolidates the changes.
  ![image](https://cloud.githubusercontent.com/assets/5641207/16981879/15181434-4e6d-11e6-8fc0-4c889d178e6f.png)
- A commit is a set of (related) changes.  You create one by editing files, calling `git add <file>` for each one you want to include, and then calling `git commit -m "meaningful message here"`.

---

Are all of your changes for your other PR already showing up on Github?  If so, you can discard your duplicate changes on your local repo.  To do that, call `git checkout -- <file>` for each file that is only relevant for your _other_ PR.  Just delete the untracked common/audio/dewinTheme folder.

Then, assuming your remaining changes are all for _this_ PR, run the following commands:

```
git stash            # hide your local, uncommitted changes for now
git pull             # get changes from your fork and apply them to your local repo
git checkout patch-3 # switch to your branch for this PR
git stash apply      # get those local changes you stashed before
```

Hopefully there are no merge conflicts.  If there are, resolve them and call `git commit`.

Now call `git diff`.  You should see all of your changes that aren't already in this PR. Call `git add <file>` for each file you changed that's relevant to this PR.  Call `git commit -m "meaningful message here"`. Call `git push`.  Then you should see your new commit here, on this PR!

Hopefully that works and I didn't miss anything. :smiley:

Commit:
Resolved merge conflicts.

Ozzystrasza commented:
@khipkin Thank you so much! I will follow that from now on, I promise.

I did everything you said. I did get merge conflicts, but I resolved them and commited. The `git diff` didn't do anything though, but that is probably cause I did `git add <files>` before that. After that I `git push` and a new commit showed here just like you said.

I made a [gist](https://gist.github.com/Ozzystrasza/3e96c6e42237d8c587e08cb6242e45f0) of everything I did, just in case.

Is there anything else I have to do now?

khipkin commented:
Looks like your local setup is good now!  However, I still don't see your tests.  Are you sure you saved them?

Ozzystrasza commented:
Awesome, one step at a time, now to the next one.
When you say my tests, do you mean the tests I made on the localhost:3000 by reproducing the issues and seeing how the changes look like and taking screen shots? If so, I edited the first comment on the PR, the one with the description, and added some images then clicked update comment. 

If it's a different kind of test I need to do, I will gladly do it! I would just need you guys to tell me what to do, again :(

crookedneighbor commented:
We're not bothering with client side tests right now as we are on the brink of a rewrite of the client side code.

khipkin commented:
@crookedneighbor ahh okay!  Good to know.

Ozzystrasza commented:
So that means it's all good now and ready for review?

Ozzystrasza committed:
Better readability

Ozzystrasza commented:
Can anyone please give me a hint about this tests I need to do? Sorry for the trouble, and thank you.

crookedneighbor commented:
We don't need test adjustments in the client code right now. Notated toward your next tier.

Ozzystrasza commented:
Thanks!
